protocol_driver_double_barrel.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "protocol_driver_double_barrel.h"
+//#include "protocol_driver_allocator.h"
+#include "absl/base/internal/sysinfo.h"
+#include "distbench_utils.h"
+#include "glog/logging.h"
+#include "google/protobuf/repeated_field.h"
+
+namespace distbench {
+
+absl::StatusOr<std::unique_ptr<ProtocolDriver>> AllocateProtocolDriver(
+    ProtocolDriverOptions opts, int* port, int tree_depth = 0);
+
+// ProtocolDriver ===============================================
+ProtocolDriverDoubleBarrel::ProtocolDriverDoubleBarrel(int tree_depth) {
+  tree_depth_ = tree_depth;
+}
+
+absl::Status ProtocolDriverDoubleBarrel::Initialize(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+
+  auto pdo = pd_opts;
+  std::string subordinate_driver_type = "";
+  google::protobuf::internal::RepeatedPtrIterator
+      <distbench::NamedSetting> iter_subordinate;
+  for (auto iter=pdo.server_settings().begin();
```

Comment:
you should be able to use range for loop here instead of begin/end

Reply:
Instead of copying every member of ProtocolDriverOptions, it is better to assign it to local variable and just remove the "next_protocol_driver". That requires the use of iterator.